Add missing index entries.

Zimmi48 opened the pull request:
This commit was originally part of #7301 but it reveals a bug in the Sphinx architecture.

cc @cpitclaudel

Zimmi48 commented on doc/sphinx/language/gallina-specification-language.rst:
```diff
@@ -217,6 +217,11 @@ numbers (see :ref:`datatypes`).
 -----
 
+.. index::
+   single: Set
```
Adding this index entry reveals a bug:
The next index entry `Set (cmd)` now appears as `(cmd)`.
On the other hand, if I put `Set (sort)` here, `Set (cmd)` is displayed normally but `Set (sort)` is displayed as `(sort)`.
I have observed the same bug with the two indices for `term` (which I added in this PR).

cpitclaudel replied:
I investigated this, and I'm not sure it's a bug :)

It's definitely a rendering issue, though; it currently looks confusing.  It comes from here: https://github.com/sphinx-doc/sphinx/blob/93837bb01ecb185d8ee64b1f794f010c510b32e1/sphinx/environment/adapters/indexentries.py#L114

We could either monkey-patch or open an issue to make this customizable.

Zimmi48 commented on doc/sphinx/language/gallina-specification-language.rst:
```diff
@@ -330,6 +343,8 @@ the type of :token:`term` to be :token:`type`.
 Inferable subterms
 ------------------
 
+.. index:: _
```
This does not appear in the symbol section of the index but in a new `_` section. Not great but not a critical problem either.

cpitclaudel replied:
Why not write this as `single: _` instead?

cpitclaudel replied:
Oh, wait, I see what you mean.  Please ignore the noise

Zimmi48 commented:
More generally, I don't really like the current situation of the index pages (cf. the general index with symbols at the beginning and at the end, the tactic index, and the error index).

maximedenes commented on doc/sphinx/language/gallina-specification-language.rst:
```diff
@@ -321,6 +330,10 @@ Section :ref:`explicit-applications`).
 
+.. index::
+   single: ... : ... (type cast)
+   single: ... <: ... (type cast)
```
Can you also add `... <<: ...`? Shouldn't we explain what those casts mean, though?

Zimmi48 replied:
Of course we should. See #7846.

maximedenes commented:
@Zimmi48 does this still need an infrastructure fix?

Zimmi48 commented:
Let me rebase this and let's evaluate what we think of the rendered result.

Zimmi48 commented:
Rendered: https://coq.gitlab.io/-/coq/-/jobs/98524916/artifacts/_install_ci/share/doc/coq/sphinx/html/genindex.html

I'm still not convinced by the rendered output of `Set` and `let ... := ...` though.

Zimmi48 commented:
I modified the PR so that indexes point to the section titles and not the text below.
I've also added a small doc of `<<:`. @maximedenes can you review again?
Rendered: https://coq.gitlab.io/-/coq/-/jobs/98558916/artifacts/_install_ci/share/doc/coq/sphinx/html/genindex.html.
I'm still not convinced by the way `Set` and `let ... := ...` are rendered but I think this PR can be merged and I'll open an issue on all the problem of the indexes I can see.
